features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/aes/aes_alt.c

```diff
@@ -21,298 +21,220 @@
  *  http://csrc.nist.gov/publications/fips/fips197/fips-197.pdf
  */
 
-#if !defined(MBEDTLS_CONFIG_FILE)
-#include "mbedtls/config.h"
-#else
-#include MBEDTLS_CONFIG_FILE
-#endif
+#include "mbedtls/aes.h"
 
 #if defined(MBEDTLS_AES_C)
 #if defined(MBEDTLS_AES_ALT)
 
 #include <string.h>
-
-#include "mbedtls/aes.h"
+#include <stdbool.h>
 
 #include "M480.h"
 #include "mbed_toolchain.h"
 #include "mbed_assert.h"
-
-
-
-#define mbedtls_trace(...) //printf(__VA_ARGS__)
+#include "mbed_error.h"
+#include "nu_bitutil.h"
+#include "crypto-misc.h"
 
 /* Implementation that should never be optimized out by the compiler */
-static void mbedtls_zeroize( void *v, size_t n ) {
-    volatile unsigned char *p = (unsigned char*)v; while( n-- ) *p++ = 0;
+static void mbedtls_zeroize( void *v, size_t n )
+{
+    volatile unsigned char *p = (unsigned char*)v;
+    while( n-- ) *p++ = 0;
 }
 
-
-static uint32_t au32MyAESIV[4] = {
-    0x00000000, 0x00000000, 0x00000000, 0x00000000
-};
-
-extern volatile int  g_AES_done;
-
-// Must be a multiple of 16 bytes block size 
+/* AES DMA compatible backup buffer if user buffer doesn't meet requirements
+ * 
+ * AES DMA buffer location requires to be:
+ * (1) Word-aligned
+ * (2) Located in 0x2xxxxxxx region. Check linker files to ensure global variables are placed in this region.
```

Comment:
@ccli8 Where is this ensured?

Reply:
@hanno-arm In `mbed-os/features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/aes/aes_alt.c` > `__nvt_aes_crypt`, DMA buffer is checked before its use:
```
if ((! crypto_dma_buff_compat(au8OutputData, MAX_DMA_CHAIN_SIZE, 16)) ||
        (! crypto_dma_buff_compat(au8InputData, MAX_DMA_CHAIN_SIZE, 16))) {
        error("Buffer for AES alter. DMA requires to be word-aligned and located in 0x20000000-0x2FFFFFFF region.");
    }
```

Reply:
@ccli8 I agree that this detects misconfiguration at runtime, but I was wondering if you could point to the relevant (parts of the) linker scripts that ensure that `au8OutputData` and `au8InputData` get a suitable address.

Reply:
@hanno-arm Take armcc scatter file `mbed-os/targets/TARGET_NUVOTON/TARGET_NUC472/device/TOOLCHAIN_ARM_STD/TARGET_NU_XRAM_SUPPORTED/NUC472.sct` for example, `au8OutputData` and `au8InputData` are static, uninitialized data, and so they would be located in `.ANY (+RW +ZI)` region, which would be within 0x20000000~0x2FFFFFFF.
<pre>
LR_IROM1 MBED_APP_START {
  ER_IROM1 MBED_APP_START {  ; load address = execution address
   *(RESET, +First)
   *(InRoot$$Sections)
   .ANY (+RO)
  }
  
  ;UVISOR AlignExpr(+0, 16) {  ; 16 byte-aligned
  ;  uvisor-lib.a (+RW +ZI)
  ;}
  
  ARM_LIB_STACK 0x20000000 EMPTY 0x800 {
  }
  
  ER_IRAMVEC 0x20000800 EMPTY (4*(16 + 142)) {  ; Reserve for vectors
  }
  
  <b>
  RW_IRAM1 AlignExpr(+0, 16) {  ; 16 byte-aligned
   .ANY (+RW +ZI)
  }
  </b>
  
  ; Too large to place into internal SRAM. So place into external SRAM instead.
  ER_XRAM1 0x60000000 {
    *lwip_* (+ZI)
    aes.o (+ZI)
    mesh_system.o (+ZI)
  }
  
  ; Extern SRAM for HEAP
  ARM_LIB_HEAP AlignExpr(+0, 16) EMPTY (0x60000000 + 0x100000 - AlignExpr(ImageLimit(ER_XRAM1), 16)) {
  }
}
</pre>